### PR TITLE
Add support for dynamic urls from tile source. Fixes #1970

### LINKED ISSUE #1970: How to replace the tile URL in a custom TileSource without disrupting the viewer?
I've been really enjoying working with OpenSeadragon!

I'm trying to figure out the best approach for a somewhat complicated feature. I'm using a custom TileSource that uses `getTileUrl` to fetch tiles from an S3 bucket. (I don't think it's relevant, but I'm also using range requests to get the tiles from a single image, so the tile URL is the same for every tile.) I'm trying to make it more secure by having the S3 signed URL expire after a fairly short time, which means I then need to update the TileSource to use this new URL. Somehow I need to seamlessly replace the URL without triggering a redraw in the viewer - it just needs to use the new URL whenever new tiles are fetched.

There are 2 approaches I'm considering:

1. Use `addTiledImage` to replace the TileSource, and somehow prevent the viewer from flickering/stuttering/freezing while the swap takes place. This seems a bit heavy handed, but probably fairly straightforward (although I haven't gotten it working yet).
2. Create a custom version of OSD that will retry a failed tile fetch (I've seen the retry logic in #1627, and would probably start there). I think I would also need to add some kind of `preRetry` function that gets called to replace the tile URL for the viewer. This seems more complicated, but possibly more robust.

In both cases, the problem boils down to finding a way to replace the URL for the TileSource without otherwise disrupting the viewer. Do you have any thoughts on the cleanest approach to accomplishing this?

Thank you for your work on this library!

### PR DESCRIPTION
**Pull request by JohnReagan**
This is a draft PR for fixing #1970 

This change updates the Tile constructor to accept either a string or function that returns a string for the `url` param. It deprecates the `.url` field in favor of a new `.getUrl()` function that checks for a function or string stored internally. References to `.url` are replaced with `.getUrl()`. 

A new comment is added to `TileSource.getTileUrl()` to suggest the new functionality. It is up to implementers of that class to use this new behavior. 

This is my first PR here, so I'd love any feedback. With this change, I aimed to be backwards compatible with the current API. I avoided using separate fields to support the static string and function getter use cases so there would be no need to check across to determine which URL is "correct." 

I'd also like feedback on what tests should be updated for this change. 

Thanks in advance for the review.

### CODE REVIEW

File: src/tile.js [hunk 1]
```diff
@@ -342,6 +344,18 @@ $.Tile.prototype = {
...
 
+    /**
+     * The URL of this tile's image.
+     * @member {String} url
+     * @memberof OpenSeadragon.Tile#
+     * @deprecated
+     * @returns {String}
+     */
+    get url() {
+        $.console.error("[Tile.url] property has been deprecated. Use [Tile.prototype.getUrl] instead.");
```
Comment (JohnReagan) [inline src/tile.js:355]: I patterned this after this PR, happy to make any changes here. https://github.com/openseadragon/openseadragon/pull/2148/files#diff-18da25a1c96bbe664a75cf01c5348ca5b54fd9fd938855d0af776d4c5b5a970dR349

**Review by iangilman — COMMENTED** [review]
This looks fantastic! The new API is consistent with the direction we've been moving for such thing, and the deprecation is perfect. I can't think of anything else that needs doing here, though if you'd like to add a test, that would be lovely. Note that our test suite currently has some known failures, so you'll have to ignore them. Anyway, testing a custom tile source that uses a getter function for its URLs should be good enough. 

Thank you for tackling this!

**Comment by JohnReagan** [comment]
Thanks for the fast review @iangilman ! I added a new test file based on this file: https://github.com/openseadragon/openseadragon/blob/master/test/modules/ajax-post-data.js

This test ensures that the viewer can handle a custom tile source that returns a function instead of a url string. Ideally, I'd have one more step to this test to change the dynamic url, trigger another ajax call, and assert it uses the new one, but I'm not quite familiar enough with the api to do that in code. Either way the core functionality is tested. Quint output below: 
image

**Comment by iangilman** [comment]
Great! The test looks good. If you want to test changing the URL, you could do: 

```
viewer.world.getItemAt(0).reset();
```

... After the first load, which will flush out the tiles and cause it to reload.

**Comment by JohnReagan** [comment]
@iangilman I didn't have any luck with the reset call to get it to re-fetch data. I did get it to re-fetch by manually deleting the existing tile from the matrix. 
`delete viewer.world.getItemAt(0).tilesMatrix[1][0][0];`
Not sure if that makes the test better or worse since we're relying on more private apis from other classes (i.e., `tilesMatix`). I'm happy to leave the test as is or push up a change with this delete call.

**Comment by iangilman** [comment]
Interesting. Sure, I think it's okay to use private APIs in the testing.

**Comment by JohnReagan** [comment]
@iangilman sounds good. I've updated the test to trigger another query with a new URL: 
image

**Review by iangilman — APPROVED** [review]
Fantastic! Thank you for putting this together :)

**Pull request merged**